src/lua/uri.c

```diff
@@ -0,0 +1,390 @@
+		}
+		break;
+	default:
+		errmsg = "Incorrect type for URI in nested table: "
+			 "should be string, number";
+		diag_set(IllegalParams, errmsg);
+		rc = -1;
+	}
+	return rc;
+}
+
+/**
+ * Create @a uri from the table, which located at the given valid @a idx.
+ */
+static int
+uri_create_from_lua_table(struct uri *uri, struct lua_State *L, int idx)
+{
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	/* There must be exactly one URI in the table */
+	int size = lua_objlen(L, idx);
+	if (size > 1 || (is_field_present(L, "uri", idx) && size == 1) ||
+	    (!is_field_present(L, "uri", idx) && size == 0)) {
+		diag_set(IllegalParams, "Incorrect URI count in nested table");
+		return -1;
+	}
+	/* Table "default_params" is not allowed for single URI */
+	if (is_field_present(L, "default_params", idx)) {
+		diag_set(IllegalParams, "Default URI query parameters are "
+			 "not allowed for single URI");
+		return -1;
+	}
+	int rc = 0;
+	if (size == 1) {
+		lua_rawgeti(L, idx, 1);
+		rc = uri_create_from_string_or_number(uri, L);
+		lua_pop(L, 1);
+	} else {
+		lua_pushstring(L, "uri");
+		lua_rawget(L, idx);
+		assert(lua_type(L, -1) != LUA_TNIL);
+		rc = uri_create_from_string_or_number(uri, L);
+		lua_pop(L, 1);
+	}
+	if (rc != 0)
+		return rc;
+	return uri_add_params_from_lua_table(uri, L, "params", idx, true);
+}
+
+/**
+ * Create @a uri_set from the table, which located at the given valid @a idx.
+ */
+static int
+uri_set_create_from_lua_table(struct uri_set *uri_set, struct lua_State *L,
+			      int idx)
+{
+	const char *errmsg;
+	int rc = 0;
+	struct uri *uri = NULL;
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	int uri_count = lua_objlen(L, idx);
+	/*
+	 * Since for replication empty table is allowed, so as not
+	 * to break backward compatibility we just return 0 here.
+	 */
+	if (uri_count == 0 && !is_field_present(L, "uri", idx))
+		return 0;
+	/*
+	 * If the number of numeric keys in the lua table is less
+	 * than or equal to one, it means that there is no more than
+	 * one URI in it.
+	 */
+	if (uri_count <= 1) {
+		uri = uri_set_add(uri_set, NULL);
+		assert(uri != NULL);
+		return uri_create_from_lua_table(uri, L, idx);
+	}
+	/*
+	 * All numeric keys corresponds to URIs in string or table
+	 * format.
+	 */
+	for (int i = 0; i < uri_count && rc == 0; i++) {
+		lua_rawgeti(L, idx, i + 1);
+		switch (lua_type(L, -1)) {
+		case LUA_TSTRING:
+		case LUA_TNUMBER:
+			if (uri_set_add(uri_set, lua_tostring(L, -1)) == NULL) {
+				errmsg = "Incorrect URI: expected host:service or "
+					 "/unix.socket";
+				diag_set(IllegalParams, errmsg);
+				rc = -1;
+			}
+			break;
+		case LUA_TTABLE:
+			uri = uri_set_add(uri_set, NULL);
+			assert(uri != NULL);
+			rc = uri_create_from_lua_table(uri, L, lua_gettop(L));
+			assert(rc == 0 || diag_get() != NULL);
+			break;
+		default:
+			errmsg = "Incorrect type for URI in table: "
+				 "should be string, number or table";
+			diag_set(IllegalParams, errmsg);
+			rc = -1;
+		}
+		lua_pop(L, 1);
```

Comment:
1. Good idea, don't see it!
2. I think it's better to have separate functions, they are too big. I specially move this code to separate functions, i think it's better to read and understand it. Also it's very ugly line breaks because of big nested of code.

Reply:
2. You could avoid nested code by removing switch-case, like this:

```C
if (this_is_a_string) {
    const char *s = get_string();
    uri_create_from_string(uri, s);
    return;
}
if (this_is_not_a_table) {
    throw error;
}
/* deal with the table */
```

Reply:
According to our talk, we will leave this comment open until the opinion of the next reviewer is clarified.